libsolidity/parsing/Parser.cpp

```diff
@@ -91,17 +91,25 @@ ASTPointer<SourceUnit> Parser::parse(CharStream& _charStream)
 	solAssert(!m_insideModifier, "");
 	try
 	{
+		m_experimentalParsingEnabledInCurrentSourceUnit = false;
 		m_recursionDepth = 0;
 		m_scanner = make_shared<Scanner>(_charStream);
 		ASTNodeFactory nodeFactory(*this);
+		bool finishedParsingPragmas = false;
```

Comment:
This is a fine way of doing this... although it would probably be simpler to just do

```suggestion
    while (m_scanner->currentToken() == Token::Pragma)
        nodes.push_back(parsePragmaDirective(false));
```

before the main loop (in which you'd make it ``parsePragmaDirective(true)``).

Maybe even reversing the boolean argument to ``parsePragmaDirective`` and calling it ``_beginningOfFile`` or so...

Reply:
That would still require leaving the `case Token::Pragma` in the main parsing loop, and I'm not exactly a fan of having this redundancy.

Reply:
Haha, well just as you said - it's cosmetics; I don't really like the boolean, but I like your suggestion even less as we'd then have two loops with two function calls. Ultimately, it's a matter of taste, and we can see what everyone else thoughts are, since if we do decide to go for your approach - it's a trivial change.

I'm more interested in the import hierarchy and how to resolve that :)

Reply:
In the very end here after this loop, we create a ``SourceUnit``. You'll probably want to store whether you've seen the pragma in that, i.e. distinguish between "experimental" and "non-experimental" source-units.

If you have the information in there as a flag, you can in ``CompilerStack::resolveImports`` easily just make sure that all ``Source``s there have this flag set in their ``ast`` members - or none of them.

Reply:
We don't need to account for separate disjoint import hierarchies that may coexist while one full hierarchy may have the pragma and another full hierarchy may not - just make sure every source file we see in one compiler run either has the pragma or not a single one has it, then we're fine and that's good enough for our purposes.